### ISSUE TITLE
How to fill in missing `parent` atttributes of mutations?

### ISSUE THREAD

**Issue by petrelharp** [tskit-dev/msprime#459]
As discussed in #122, we decided to add a `parent` property of mutations to make things more efficient. It is also necessary if there is more than one mutation on a given branch. Since this property is encoded by the trees themselves (except in the latter case), we wanted to provide a method to fill in this column for tables where it isn't provided already: #286 and #288.

However:

1. computing the parent column of the mutation table requires constructing the trees and iterating through them, but
2. constructing a proper tree sequence requires the parent column (at least, to get genotypes correctly; without this, mutations may be in the wrong order within each site).

The proposed fix was to add a flag to `load_tables` that would go through and correctly set the parents of each mutation. We have code for this that will appear in #286 soon thanks to @bhaller. Then, we could dump the tables, and read them back in again... but it seems like there should be a better way to do this?

By the way, the flag to `load_tables` (or, maybe a different flag?) can also *check* the parent column (basically, compute the parents; fill in if they are `MSP_NULL_NODE` and throw an error if they exist but are wrong).  This is a possibly important check we haven't got in yet.

**Comment by jeromekelleher** [comment]
This seems like an excellent plan to me; I'm looking forward to seeing the code!

**Comment by petrelharp** [comment]
Update:

1. Checking when there's more than one mutation per branch is a little tricky (but do-able).
2. The previous code (e.g., python) assumes the mutations come in the right order, which they don't necessarily; this needs to be fixed to produce (and, check) parents correctly even if they aren't ordered correctly.

Here, "ordered correctly" means ordered so that any mutation comes after its parent; an easy way to ensure this is to order by (-1) * node time.  Currently `sort_tables` sorts mutations by site, and then by ID, thus preserving ordering within each site.  If mutations are input in order by time, then sort_tables should thus preserve time ordering.  However, it turns out that `simplify_tables` does not always preserve this ordering, at least if mutation parents are not set (sorry, no MWE yet). 

This means that what we want to happen is:

1. Produce tables in forwards time, with mutations ordered by time of appearance, and mostly leaving out mutation.parent information (only providing it when there's more than one mutation on actually the same node).
2. `sort`, `clean`, and `simplify` tables. 
3. At the end, when we want a tree sequence, fill in the `mutation.parent` information, by building a tree sequence and working out whot the parents are.

However, what actually happens is:

3. `simplify`, lacking `mutation.parent` information, has make the mutations out of order. We can still build `mutation.parent` correctly, but to get it back into a tree sequence we have to write out the tables, *reorder mutations*, and read them in again.

Currently, neither `sort_tables` or `simplify_tables` notices if a mutation appears in the table before its parent (even if `mutation.parent` is specified).  So, if in step (3.) we added `mutation.parent`, wrote out the tables, put them through `sort_tables`and read them in again, we'd get an `Parent mutation ID must be < current ID.` error.

So, there are two issues:

1. `simplify_tables` is not preserving relative order of mutations at the same site.  Maybe the new version of `simplify` is, though, I haven't checked.  If it *did*, then our workflow would be all good, and the `check_parents` flag will prevent us from reading in any inconsistent tables.

2. `sort_tables`, provided with a correct `mutation.parent` column, has enough information to provide valid tables in which mutations come after their parents; but it currently does not. If we can fix (1) then no-one actually needs this functionality, but it's an annoying caveat.

Next I'll be looking at `simplify` to see why it's unordering mutations.

**Comment by petrelharp** [comment]
Two more notes:

1. In calculating parents, we are sorting mutations by `node` also. Perhaps this should just be added to `sort_tables` (and the requirements for valid tables).

2. I think the following example shows that there's no simple key that will let us use `qsort` to sort the tables so that mutations always come after their parent mutations:
```
id  node derived_state parent
0   0              1                1
1   0              1               -1
2   0              0                0
```
The only thing to sort on is `parent`, but using it gives the wrong ordering.  So, if `sort_tables` is to correctly order this table, it'd have to (a) sort on node time, and then (b) for each set of mutations occurring on the same node, go find the linear ordering of them within that node.

**Comment by jeromekelleher** [comment]
> simplify_tables is not preserving relative order of mutations at the same site. Maybe the new version of simplify is, though, I haven't checked. If it did, then our workflow would be all good, and the check_parents flag will prevent us from reading in any inconsistent tables.

Sounds good @petrelharp --- this seems like a bug to me and we should fix it. One thing that's bitten me repeatedly is that glibc's ``qsort`` is stable but other platforms are not. So, unless you complete specify your sort keys things will seem to work fine on Linux, but will break unexpectedly on macs. Just something to bear in mind as a possible source of simplify's "instabililty".

**Comment by petrelharp** [comment]
Update: turns out the problem was *not* in simplify - I checked, and `simplify_tables` should definately preserve ordering within the mutation tables - it was occurring because

1. we had more than one site at the same position when calling sort tables
2. `sort_tables` was switching the order of those sites (maybe only on the Mac)
3. and therefore switching the order of the mutations that referred to those two sites (which were actually the same)
4. and then simpliyfing
5. and THEN we were calling `deduplicate_sites`

This could have been fixed by calling `deduplicate_sites` before sort & simplify... but we wanted to deduplicate the smaller site table.. and @bhaller has a solution, which is to add `site.id` to the keys that `sort_tables` sorts on, thus preserving order within sites having the same position.

PR forthcoming.

**Comment by jeromekelleher** [comment]
> This could have been fixed by calling deduplicate_sites before sort & simplify... but we wanted to deduplicate the smaller site table.. and @bhaller has a solution, which is to add site.id to the keys that sort_tables sorts on, thus preserving order within sites having the same position.

Excellent, I entirely approve of this solution!

**Comment by petrelharp** [comment]
Closed in #461
